src/vs/editor/standalone/browser/standaloneLanguages.ts

```diff
@@ -553,7 +570,7 @@ export interface CompletionItemProvider {
 	/**
 	 * Provide completion items for the given position and document.
 	 */
-	provideCompletionItems(model: editorCommon.IReadOnlyModel, position: Position, token: CancellationToken): CompletionItem[] | Thenable<CompletionItem[]> | CompletionList | Thenable<CompletionList>;
+	provideCompletionItems: ProviderCompletionItems | ProviderCompletionItemsForContext;
```

Comment:
I see how this makes the implementation nicer but I think this will add confusion... I'd go simple and just add another overload of the `provideCompletionItems`-function, maybe deprecate the old one

Reply:
I tried this but ran into some issues with the type system. Here's a simple example of the problem:

```
interface X { x: number }

interface Y { y: number }

interface IFoo {
  a(x: X): string
  a(x: Y, y: boolean): string
}


const z: IFoo = { 
  a(x: X) { return '' }
} 

if (z.a.length <= 1) {
  z.a(1)
}
```

Will complain that `property x is missing on type Y`. I also tried a few other solutions but haven't found how to express this yet in a backwards compatible way. Here's an issue I opened related to this: https://github.com/Microsoft/TypeScript/issues/18422 

Let me know if you have any suggestions. We could move these types into a namespace to hide them better perhaps? Otherwise this signature sort of works:

```

		provideCompletionItems: ((document: TextDocument, position: Position, token: CancellationToken)=> ProviderResult<CompletionItem[] | CompletionList>) |
		((document: TextDocument, position: Position, context: CompletionContext, token: CancellationToken)=> ProviderResult<CompletionItem[] | CompletionList>);
```

However it cannot really be documented well and is also pretty confusing for extensions authors when dealing with  intellisense

Reply:
Hm, bummer... I'll give it some thought but it might be like that...